RecoEgamma/EgammaTools/src/ElectronEnergyCalibrator.cc

```diff
@@ -0,0 +1,208 @@
+						   const std::string& correctionFile) :
+  correctionRetriever_(correctionFile), 
+  epCombinationTool_(&combinator), 
+  rng_(nullptr),
+  minEt_(1.0)
+{
+  
+}
+
+void ElectronEnergyCalibrator::initPrivateRng(TRandom *rnd)
+{
+  rng_ = rnd;
+}
+
+std::vector<float> ElectronEnergyCalibrator::
+calibrate(reco::GsfElectron &ele,
+	  const unsigned int runNumber, 
+	  const EcalRecHitCollection *recHits, 
+	  edm::StreamID const & id, 
+	  const ElectronEnergyCalibrator::EventType eventType) const
+{
+  return calibrate(ele,runNumber,recHits,gauss(id),eventType);
+}
+
+std::vector<float> ElectronEnergyCalibrator::
+calibrate(reco::GsfElectron &ele, unsigned int runNumber, 
+	  const EcalRecHitCollection *recHits, 
+	  const float smearNrSigma, 
+	  const ElectronEnergyCalibrator::EventType eventType) const
+{
+  const float scEtaAbs = std::abs(ele.superCluster()->eta());
+  const float et = ele.ecalEnergy() / cosh(scEtaAbs);
+
+  if (et < minEt_) {
+    std::vector<float> retVal(EGEnergySysIndex::kNrSysErrs,ele.energy());
+    retVal[EGEnergySysIndex::kScaleValue]  = 1.0;
+    retVal[EGEnergySysIndex::kSmearValue]  = 0.0;
+    retVal[EGEnergySysIndex::kSmearNrSigma] = smearNrSigma;
+    retVal[EGEnergySysIndex::kEcalPreCorr] = ele.ecalEnergy();
+    retVal[EGEnergySysIndex::kEcalErrPreCorr] = ele.ecalEnergyError();
+    retVal[EGEnergySysIndex::kEcalPostCorr] = ele.ecalEnergy();
+    retVal[EGEnergySysIndex::kEcalErrPostCorr] = ele.ecalEnergyError();
+    retVal[EGEnergySysIndex::kEcalTrkPreCorr] = ele.energy();
+    retVal[EGEnergySysIndex::kEcalTrkErrPreCorr] = ele.corrections().combinedP4Error;
+    retVal[EGEnergySysIndex::kEcalTrkPostCorr] = ele.energy();
+    retVal[EGEnergySysIndex::kEcalTrkErrPostCorr] = ele.corrections().combinedP4Error;
+    return retVal;
+  }
+
+  const DetId seedDetId = ele.superCluster()->seed()->seed();
+  EcalRecHitCollection::const_iterator seedRecHit = recHits->find(seedDetId);
+  unsigned int gainSeedSC = 12;
+  if (seedRecHit != recHits->end()) { 
+    if(seedRecHit->checkFlag(EcalRecHit::kHasSwitchToGain6)) gainSeedSC = 6;
+    if(seedRecHit->checkFlag(EcalRecHit::kHasSwitchToGain1)) gainSeedSC = 1;
+  }
+
+  const EnergyScaleCorrection::ScaleCorrection* scaleCorr = correctionRetriever_.getScaleCorr(runNumber, et, scEtaAbs, ele.full5x5_r9(), gainSeedSC);  
+  const EnergyScaleCorrection::SmearCorrection* smearCorr = correctionRetriever_.getSmearCorr(runNumber, et, scEtaAbs, ele.full5x5_r9(), gainSeedSC);  
+  if(scaleCorr==nullptr) scaleCorr=&defaultScaleCorr_;
+  if(smearCorr==nullptr) smearCorr=&defaultSmearCorr_;
+  
+  std::vector<float> uncertainties(EGEnergySysIndex::kNrSysErrs,0.);
+  
+  uncertainties[EGEnergySysIndex::kScaleValue]  = scaleCorr->scale();
+  uncertainties[EGEnergySysIndex::kSmearValue]  = smearCorr->sigma(et); //even though we use scale = 1.0, we still store the value returned for MC
+  uncertainties[EGEnergySysIndex::kSmearNrSigma]  = smearNrSigma;
```

Comment:
Its intentional. Data records this value (because why not) but does not apply it.  Not particularly married to it, can zero it out.  Will have a think.

Reply:
of course I'll have to properly record it (and for the MC scale) if we decide to implement the above comment on scaling optionally applying to MC and smearing optionally applying to data.